x-wing/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+#![cfg_attr(not(test), no_std)]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![doc = include_str!("../README.md")]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo.svg"
+)]
+
+//! # Usage
+//!
+//! This crate implements the X-Wing Key Encapsulation Method (X-Wing-KEM) algorithm.
+//! X-Wing-KEM is a KEM in the sense that it creates an (decapsulation key, encapsulation key) pair,
+//! such that anyone can use the encapsulation key to establish a shared key with the holder of the
+//! decapsulation key. X-Wing-KEM is a general-purpose hybrid post-quantum KEM, combining x25519 and ML-KEM-768.
+//!
+//! ```
+//! # use rand;
+//! # use kem::{Decapsulate, Encapsulate};
+//! let mut rng = rand::thread_rng();
+//!
+//! let (sk, pk) = x_wing::generate_key_pair(&mut rng);
+//!
+//! let (ct, ss_sender) = pk.encapsulate(&mut rng).unwrap();
+//!
+//! let ss_receiver = sk.decapsulate(&ct).unwrap();
+//!
+//! assert_eq!(ss_sender, ss_receiver);
+//! ```
+//!
+//! [X-Wing draft]: https://datatracker.ietf.org/doc/html/draft-connolly-cfrg-xwing-kem
+
+use kem::{Decapsulate, Encapsulate};
+use ml_kem::array::ArrayN;
+use ml_kem::{kem, EncodedSizeUser, KemCore, MlKem768, MlKem768Params, B32};
+use rand_core::CryptoRngCore;
+use sha3::digest::core_api::XofReaderCoreWrapper;
+use sha3::digest::{ExtendableOutput, XofReader};
+use sha3::{Sha3_256, Shake128, Shake128ReaderCore};
+use x25519_dalek::{x25519, X25519_BASEPOINT_BYTES};
+#[cfg(feature = "zeroize")]
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+type MlKem768DecapsulationKey = kem::DecapsulationKey<MlKem768Params>;
+type MlKem768EncapsulationKey = kem::EncapsulationKey<MlKem768Params>;
+
+const X_WING_LABEL: &[u8; 6] = br"\.//^\";
+
+const PUBLIC_KEY_SIZE: usize = 1216;
+const PRIVATE_KEY_SIZE: usize = 32;
+
+/// Shared secret key
+pub type SharedSecret = [u8; 32];
+
+// The naming convention of variables matches the RFC.
+// ss -> Shared Secret
+// ct -> Cipher Text
+// ek -> Ephemeral Key
+// pk -> Public Key
+// sk -> Secret Key
+// Postfixes:
+// _m -> ML-Kem related key
+// _x -> x25519 related key
+
+/// The X-Wing encapsulation or public key
+#[derive(Clone, PartialEq)]
+pub struct EncapsulationKey {
+    pk_m: MlKem768EncapsulationKey,
+    pk_x: x25519_dalek::PublicKey,
+}
```

Comment:
For some reason `kem::EncapsulationKey<MlKem768Params>` does not implement `Eq`, as far as I can tell implementing `Eq` should not be a problem as all underling types are u8 arrays or integers.

Reply:
You can add a derive for it